### ISSUE TITLE
[bug] It's impossible to set winsdk_version argument for vcvars script

### ISSUE THREAD

**Issue by apyvovar** [conan-io/conan#14383]
### Environment details

* Windows 11
* MSVC 2022
* Conan version: 2.0.9
* Python version: 3.10

### Steps to reproduce

The generated `conanvcvars.bat` contains the following line

```
set "VSCMD_START_DIR=%CD%" && call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC/Auxiliary/Build/vcvarsall.bat"  amd64 -vcvars_ver=14.3
```

`winsdk_version` is not set at all.
For some legacy code I need to set older SDK version for MSVC among several installed.
Please, correct me if I wrong, but I looked through the documentation and python scripts and it's looks like it's impossible to set it.

Thank you.

**Comment by memsharded** [comment]
Hi @apyvovar 

Yes, you are right, at the moment there is no way to set the ``winsdk_version``.

It shouldn't be difficult to define a way, but I am not sure what would be the right mechanism to input it:
- Is it a ``conf`` like ``tools.microsoft:winsdk_version`` that the ``VCVars`` generator reads it? Or is it an optional ``settings.yml`` subsetting for the ``os=Windows``?
- What other generators need to received it? Is it ``CMAKE_SYSTEM_VERSION`` for cmake, or better ``-T v143,version=14.31.31103``?
- How MSBuild, Meson or Autotools could define this?

**Comment by apyvovar** [comment]
He, @memsharded 

IMO, since we already have `tools.microsoft.msbuild:vs_version` section in `conf`, it's seems natural to add `winsdk_version` there as well.

Unfortunately, I didn't have change to check it on other generators, so I can't say the way to implement it them.

Hope, the issue will be resolved soon.

Thank you.

**Comment by memsharded** [comment]
> IMO, since we already have tools.microsoft.msbuild:vs_version

This is the discussion. We know that the ``vs_version`` given you use the same toolset won't produce different binaries. So it is important to know if the ``winsdk_version`` could affect somehow and produce different binaries. Apparently that this is not the case (can @jcar87 confirm?) and the ``winsdk_version`` could be a ``conf``.

``Meson``, ``Autotools`` and ``MSBuild`` Toolchains will be ok, because they use ``VCVars`` under the hood too. But ``CMakeToolchain`` we really need to know how to do it before proceeding, this is a must if we want to support this new conf.

**Comment by MarcelRaad** [comment]
The Windows SDK version can produce different binaries. I'm not sure if on its own, but there are libraries that check preprocessor macros from the Windows SDK and include different code depending on their existence/value. Therefore, we currently have that in `options` (`settings` would make more sense though, as already done for the Apple SDKs).

We currently (with Conan v1.60.2) use it for the CMake and MSBuild/VisualStudio build helpers. For CMake versions up to 3.26, the setting is `CMAKE_SYSTEM_VERSION`, but that changed with https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8389 (I haven't updated to 3.27 yet, so I'm not familiar with the new way).

**Comment by jcar87** [comment]
> The Windows SDK version can produce different binaries. I'm not sure if on its own, but there are libraries that check preprocessor macros from the Windows SDK and include different code depending on their existence/value. Therefore, we currently have that in `options` (`settings` would make more sense though, as already done for the Apple SDKs).

This appears to be unclear from the microsoft documentation however, it appear that as you say, the keyword is "on its own" - the current windows 10 SDKs are compatible with Windows 7 SP1 onwards, as per https://learn.microsoft.com/en-us/cpp/windows/how-to-use-the-windows-10-sdk-in-a-windows-desktop-application?view=msvc-170


According to this [stackoverflow](https://stackoverflow.com/questions/39621646/what-does-the-target-platform-version-mean-for-a-vs-c-project/56084339#56084339), in Visual Studio 2019, the "Target Platform Version" field in the C++ projects was renamed to "Windows SDK Version". I would guess this is a good thing since otherwise it seems confusing, and may imply that binaries will only work with that version of Windows.



When features from specific versions are required, one can 'target" newer versions of Windows:
> For more information about targeting specific versions of Windows, see [Using the Windows Headers](https://learn.microsoft.com/en-us/windows/win32/WinProg/using-the-windows-headers) and [Update WINVER and _WIN32_WINNT](https://learn.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt?view=msvc-170).

This is indeed done with macros.  

There's some additional information here: https://walbourn.github.io/a-brief-history-of-windows-sdks/  

What I take from all this:
* Using a recent version of the SDK doesn't necessarily make binaries incompatible with older versions of Windows, this depends on the actual API calls made by the code. Visual Studio, now CMake (as per the recent 3.27 update) and vcvars all choose the most recent version of the SDK available, so I think that's a strong indicator that this is not a big issue and using a recent version is recommended. 
* The SDKs support a range of versions, with the lower bound being Windows 7 SP1 and the upper bound being some arbitrary Windows 10/Windows 11 version (presumably the one matching the version of the SDK itself).

**Comment by memsharded** [comment]
Lets move this forward:
- New conf to define it
- Users that want to make it part of the ``package_id`` can do it via other ``conf``
- Vcvars itnegration is easy
- @jcar87 will help with CMake integration
- @memsharded will check MSBuildToolchain integration

**Comment by hekrdla** [comment]
Hi, I'm looking for this support to. I can give some inputs, since I already spend some time on debugging:

- Conan settings : apple platforms in settings.yml already support `os.sdk_version`, what about to add same for Windows
- CMake: 
    - sdk is selected based on this conditions: [CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION](https://cmake.org/cmake/help/latest/variable/CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION.html) implemented here [cmGlobalVisualStudio14Generator.cxx](https://github.com/Kitware/CMake/blob/master/Source/cmGlobalVisualStudio14Generator.cxx#L142)
    - [CMAKE_SYSTEM_VERSION](https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_VERSION.html#variable:CMAKE_SYSTEM_VERSION) is important
        - for x86/x86_64 CMAKE_SYSTEM_VERSION defaults to CMAKE_HOST_SYSTEM_VERSION aka conan build context
        - for arm/arm64 (Cross Compiling) has no default, in this case, a lot of CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION conditions are not valid

Update: @memsharded 
- Looks like for cmaketoolchain this is already supported in Conan 1.X with [tools.cmake.cmaketoolchain:system_version](https://github.com/conan-io/conan/blob/release/2.0/conan/tools/cmake/toolchain/blocks.py#L744) when cross building
- And for WindowsStore/WindowsCE, settings.yml already support `os.version: ["8.1", "10.0"]`/`os.version: ["5.0", "6.0", "7.0", "8.0"]
` which will be used for [CMAKE_SYSTEM_VERSION](https://github.com/conan-io/conan/blob/release/2.0/conan/tools/cmake/toolchain/blocks.py#L764) when cross building. So just updating settings.yml for Windows can achieve same
- If this would not be only for cross building, it can be used for x86/x86_64 on windows
- also documentation not mention that `tools.cmake.cmaketoolchain:system_...` are only for cross build

Note: In my case, I try to build windows arm64 and since I have both winsdk 8.1 and 10.0 installed and CMAKE_SYSTEM_VERSION is not set, CMake choose 8.1 :-(

Can this be ported also into Conan 1.X?

**Comment by memsharded** [comment]
Proposing initial draft PR in https://github.com/conan-io/conan/pull/15272

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15272 has been merged, for next 2.0.15, feedback welcome
